docs/architecture-and-design.md

```diff
@@ -0,0 +1,390 @@
+4. Being a defender that tries to steal the ball from enemies
+
+Tactics use [Actions](#skills--actions) to implement their behavior. Using the [Action](#skills--actions) abstraction makes it much easier for Tactics to express what they want to do, and make it easier to design and implement behavior. Tactics can focus more on what things to do, and when to do them, in order to implement more intelligent and adaptable behavior.
+
+### Plays
+The `P` in `STP` stands for `Plays`. A `Play` represents a "team-wide goal" for the robots. They can be thought of much like Plays in real-life soccer. Examples include:
+1. A Play for taking friendly corner kicks
+2. A Play for defending enemy kickoffs
+3. A general defense play
+4. A passing-based offense play
+5. A dribbling-based offense play
+
+Plays are made up of `Tactics`. Plays can have "stages" and change what `Tactics` are being used as the state of the game changes, which allows us to implement more complex behavior. Read the section on [Coroutines](#coroutines) to learn more about how we write strategy with "stages".
+
+Furthermore, every play specifies an `Applicable` and `Invariant` condition. These are used to determine what plays should be run at what time, and when a Play should terminate.
+
+`Applicable` indicates when a `Play` can be started. For example, we would not want to start a `Defense Play` if our team is in possession of the ball. The `Invariant` condition is a condition that must always be met for the `Play` to continue running. If this condition ever becomes false, the current `Play` will stop running and a new one will be chosen. For example, once we start running a friendly `Corner Kick` play, we want the `Play` to continue running as long as the enemy team does not have possession of the bali.
+
+
+## Navigation
+The `Navigator` is responsible for path planning and navigation. Once our strategy has decided what it wants to do, it passes the resulting [Intents](#intents) to the `Navigator`. The `Navigator` is then responsible for breaking down the [Intents](#intents) and turning them into [Primitives](#primitives).
+
+Most [Intents](#intents) are easy to break down into  [Primitives](#primitives), and can typically just be converted directly without having to do any extra work. However, some [Intents](#intents) like the `MoveIntent` rely on the navigator to implement more complex behavior like obstacle avoidance. This is where the "Navigation" part of the `Navigator` comes in.
+
+In order for a robot to move to the desired destination of a `MoveIntent`, the Navigator will use various path-planning algorithms to find a path across the field that does not collide with any robots or violate any restrictions set on the `MoveIntent`. The Navigator then translates this path into a series of `MovePrimitives`, which are sent to the robot sequentially so that it follows the planned path across the field.
+
+## AI Diagram
+![Backend Diagram](images/ai_diagram.svg)
+
+
+# Visualizer
```

Comment:
From the ticket, it seems that we want a bit more detail in visualizer (i.e. Lambdas). I was wondering whether we should add a bit more detail to the visualizer?

Reply:
Yes more detail should be added. At this point I'll do that in a separate PR once a few more changes are made to the Visualizer. I'll mark #880 as not resolved by this PR since you're right I didn't address most of its points.